### DISCUSSION TITLE
Diagnostics + Git Diff + Filename

### DISCUSSION THREAD

**Discussion by lkhphuc** [b0o/incline.nvim#32] [Showcase]
![Screenshot 2022-11-09 at 14 27 18](https://user-images.githubusercontent.com/12573521/200856241-d936bd21-bdb3-4348-9108-94fc72d4f2de.png)

Inspired by  #31 , I copied the code and added gitsign integration. 
I still use lualine but this obviates my need for tabline.
Cheers.

```lua
require('incline').setup({
     render = function(props)
        local filename = vim.fn.fnamemodify(vim.api.nvim_buf_get_name(props.buf), ":t")
        local ft_icon, ft_color = require("nvim-web-devicons").get_icon_color(filename)
        local modified = vim.bo[props.buf].modified and 'bold,italic' or 'bold'

        local function get_git_diff()
          local icons = { removed = "", changed = "", added = "" }
          icons["changed"] = icons.modified
          local signs = vim.b[props.buf].gitsigns_status_dict
          local labels = {}
          if signs == nil then return labels end
          for name, icon in pairs(icons) do
            if tonumber(signs[name]) and signs[name] > 0 then
              table.insert(labels, { icon .. signs[name] .. " ", group = "Diff" .. name })
            end
          end
          if #labels > 0 then table.insert(labels, { "┊ " }) end
          return labels
        end
        local function get_diagnostic_label()
          local icons = { error = '', warn = '', info = '', hint = '', }
          local label = {}

          for severity, icon in pairs(icons) do
            local n = #vim.diagnostic.get(
              props.buf,
              { severity = vim.diagnostic.severity[string.upper(severity)] }
            )
            if n > 0 then
              table.insert(label, { icon .. n .. " ", group = "DiagnosticSign" .. severity })
            end
          end
          if #label > 0 then table.insert(label, { "┊ " }) end
          return label
        end

        local buffer = {
          { get_diagnostic_label() },
          { get_git_diff() },
          { (ft_icon or "") .. " ", guifg = ft_color, guibg = "none" },
          { filename .. " ", gui = modified },
          { "┊  " .. vim.api.nvim_win_get_number(props.win), group = "DevIconWindows" },
        }
        return buffer
    end,

})
```

**Comment by b0o** [comment]
Very nice! Love your color scheme, too. Thank you for sharing.

**Comment by alandu12** [comment]
Really nice share! Thanks bro.

**Comment by b0o** [comment]
Hey @lkhphuc, is it okay with you for me to add this as an example in the README, along with the screenshot?

**Reply by lkhphuc** [reply]
Yes please feel free to. I've been using this for year and really happy with it instead of bufferline. The only thing I added was window number indicator for quick switching, i.e `2<C-w><C-w>`

```
        local buffer = {
          { get_diagnostic_label() },
          { get_git_diff() },
          { ft_icon .. " ", guifg = ft_color, guibg = "none" },
          { filename .. " ", gui = modified },
          {"┊ "}, { " " .. vim.api.nvim_win_get_number(props.win), group = "DevIconWindows" },
        }

```

<img width="563" alt="Screenshot 2024-01-24 at 12 42 25" src="https://github.com/b0o/incline.nvim/assets/12573521/7d38f952-a687-46c1-94a8-0d1b01299703">

**Comment by LudoPinelli** [comment]
> ```lua
> require('incline').setup({
>   render = function(props)
> 
>     local filename = vim.fn.fnamemodify(vim.api.nvim_buf_get_name(props.buf), ':t')
>     local ft_icon, ft_color = require("nvim-web-devicons").get_icon_color(filename)
>     local modified = vim.api.nvim_buf_get_option(props.buf, "modified") and "bold,italic" or "bold"
> [...]
>   end
> })
> ```

There is a slight change to be made here to maintain the code compatible with next nvim versions, `nvim_buf_get_option()` is deprecated, `nvim_get_option_value()` should be used instead:
```lua
    local modified = vim.api.nvim_get_option_value("modified", { buf = props.buf })
    and "bold,italic" or "bold"

```

P.S: very nice plugin! And very nice config for it!

**Reply by b0o** [reply]
You can also use this:

```lua
local modified = vim.bo[props.buf].modified
```

**Reply by lkhphuc** [reply]
> local modified = vim.bo[props.buf].modified

Actually this does not work but I am not sure why. 
I updated the original post with new nvim's API anyway.

**Reply by b0o** [reply]
Really? Works for me (I left out the second part of the line in my comment above):

```lua
local modified = vim.bo[props.buf].modified and 'bold,italic' or 'bold'
```

**Reply by lkhphuc** [reply]
Ah no my bad. It works perfectly now.

**Comment by egoarka** [comment]
In case someone is looking for `coc.nvim` diagnostics, use this snippet.


```lua
local function get_diagnostic_label()
  local icons = { { "error", '' }, { "warn", '' }, { "info", '' }, { "hint", '' } }
  local label = {}

  local ok, diag = pcall(vim.api.nvim_buf_get_var, props.buf, "coc_diagnostic_info")
  if not ok then return label end

  for i, count in ipairs({ diag.error, diag.warning, diag.information, diag.hint }) do
    if count > 0 then
      local severity, icon = unpack(icons[i])
      table.insert(label, { icon .. " " .. count .. " ", group = "DiagnosticSign" .. severity })
    end
  end

  if #label > 0 then table.insert(label, { "┊ " }) end
  return label
end

```

**Comment by ilan-schemoul** [comment]
On windows terminal (wsl) when I typed "p" to paste it converted into "?" (literally this symbol). To make it work I needed to do in insert mode ctrl+shift+v

**Comment by ilan-schemoul** [comment]
```
          local icons = { removed = "", changed = "", added = "" }
          icons["changed"] = icons.modified
```
I guess you meant 
`          local icons = { removed = "", modified= "", added = "" }`

**Comment by YousefHadder** [comment]
for some reason, I'm missing the colors for the removed/changed/added icons

<img width="315" height="30" alt="Screenshot 2025-08-06 at 1 54 39 PM" src="https://github.com/user-attachments/assets/f2eae672-e2d8-42df-bc9c-0b15eb347ffd" />
